catalog/transformer/catalog-transformer-zip/src/main/java/org/codice/ddf/catalog/transformer/zip/ZipCompression.java

```diff
@@ -13,261 +13,159 @@
+    String transformerId = arguments.getOrDefault(TRANSFORMER_ID, "").toString();
 
-      // write the metacards to the zip
-      resultList
-          .stream()
-          .map(Result::getMetacard)
-          .forEach(
-              metacard -> {
-                writeMetacardToZip(zipOutputStream, metacard);
-
-                if (hasLocalResources(metacard)) {
-                  resourceMap.putAll(getAllMetacardContent(metacard));
-                }
-              });
+    if (StringUtils.isBlank(transformerId)) {
+      throw new CatalogTransformerException("A valid transformer ID must be provided.");
+    }
 
-      // write the resources to the zip
-      resourceMap.forEach(
-          (filename, resource) -> writeResourceToZip(zipOutputStream, filename, resource));
+    InputStream inputStream = createZip(sourceResponse, transformerId);
 
-    } catch (IOException e) {
-      throw new CatalogTransformerException(
-          String.format(
-              "Error occurred when initializing/closing ZipOutputStream with path %s.", filePath),
-          e);
-    }
+    return new BinaryContentImpl(inputStream, mimeType);
   }
 
-  private void writeMetacardToZip(ZipOutputStream zipOutputStream, Metacard metacard) {
+  private InputStream createZip(SourceResponse sourceResponse, String transformerId)
+      throws CatalogTransformerException {
+    ServiceReference<MetacardTransformer> serviceRef =
+        getTransformerServiceReference(transformerId);
+    MetacardTransformer transformer = bundleContext.getService(serviceRef);
 
-    try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-        ObjectOutputStream objectOutputStream = new ObjectOutputStream(byteArrayOutputStream)) {
+    String extension = getFileExtensionFromService(serviceRef);
 
-      ZipEntry zipEntry = new ZipEntry(METACARD_PATH + metacard.getId());
-      zipOutputStream.putNextEntry(zipEntry);
+    if (StringUtils.isNotBlank(extension)) {
+      extension = "." + extension;
+    }
 
-      objectOutputStream.writeObject(new MetacardImpl(metacard));
-      InputStream inputStream = new ByteArrayInputStream(byteArrayOutputStream.toByteArray());
+    try (FileBackedOutputStream fileBackedOutputStream = new FileBackedOutputStream(BUFFER_SIZE);
+        ZipOutputStream zipOutputStream = new ZipOutputStream(fileBackedOutputStream)) {
 
-      IOUtils.copy(inputStream, zipOutputStream);
-      inputStream.close();
-    } catch (IOException e) {
-      LOGGER.debug("Failed to add metacard with id {}.", metacard.getId(), e);
-    }
-  }
+      for (Result result : sourceResponse.getResults()) {
+        Metacard metacard = result.getMetacard();
 
-  private boolean hasLocalResources(Metacard metacard) {
-    URI uri = metacard.getResourceURI();
-    return (uri != null && ContentItem.CONTENT_SCHEME.equals(uri.getScheme()));
-  }
+        BinaryContent binaryContent = transformer.transform(metacard, Collections.emptyMap());
```

Comment:
❓ If a single transform fails, does the entire operation fail?  Seems like we'd want to do best effort here.